matlab/src/feather_writer.cc

```diff
@@ -0,0 +1,365 @@
+    mexErrMsgIdAndTxt("MATLAB:arrow:MissingStructField",
+                      "Missing field '%s' in input struct array", fieldname);
+  }
+
+  mxClassID actual_class_id = mxGetClassID(field);
+
+  // Avoid type check if an mxUNKNOWN_CLASS is provided since the UNKNOWN type is used to
+  // signify genericity in the input type.
+  if (expected_class_id != mxUNKNOWN_CLASS) {
+    if (expected_class_id != actual_class_id) {
+      mexErrMsgIdAndTxt("MATLAB:arrow:MissingStructField",
+                        "Incorrect type '%s' for struct array field '%s'",
+                        mxGetClassName(field), fieldname);
+    }
+  }
+
+  // Some struct fields (like the table description) can be empty, while others 
+  // (like NumRows) should never be empty. This conditional helps account for both cases.
+  if (!can_be_empty) {
+    // Ensure that individual mxStructArray fields are non-empty.
+    // We can call mxGetData after this without needing another null check.
+    if (mxIsEmpty(field)) {
+      mexErrMsgIdAndTxt("MATLAB:arrow:EmptyStructField",
+                        "Struct array field '%s' cannot be empty", fieldname);
+    }
+  }
+}
+
+// Utility function to convert mxChar mxArray* to std::string while preserving
+// Unicode code points.
+std::string MxArrayToString(const mxArray* array) {
+  // Return empty std::string if a mxChar array is not passed in.
+  if (!mxIsChar(array)) {
+    return std::string();
+  }
+
+  // Convert mxArray first to a C-style char array, then copy into a std::string.
+  char* utf8_array = mxArrayToUTF8String(array);
+  std::string output(utf8_array);
+
+  // Free the allocated char* from the MEX runtime.
+  mxFree(utf8_array);
+
+  return output;
+}
+
+// Compare number of columns and exit out to the MATLAB layer if incorrect.
+void ValidateNumColumns(int64_t actual, int64_t expected) {
+  if (actual != expected) {
+    mexErrMsgIdAndTxt("MATLAB:arrow:IncorrectNumberOfColumns",
+                      "Received only '%d' columns but expected '%d' columns", actual,
+                      expected);
+  }
+}
+
+// Compare number of rows and exit out to the MATLAB layer if incorrect.
+void ValidateNumRows(int64_t actual, int64_t expected) {
+  if (actual != expected) {
+    mexErrMsgIdAndTxt("MATLAB:arrow:IncorrectNumberOfRows",
+                      "Received only '%d' rows but expected '%d' rows", actual, expected);
+  }
+}
+
+// Calculate the number of bytes required in the bit-packed validity buffer.
+constexpr int64_t BitPackedLength(int64_t num_elements) {
+  // Since mxLogicalArray encodes [0, 1] in a full byte, we can compress that byte
+  // down to a bit...therefore dividing the mxLogicalArray length by 8 here.
+  return static_cast<int64_t>(std::ceil(num_elements / 8.0));
+}
+
+// Construct a bit-packed buffer on the heap for validity information.
+std::shared_ptr<uint8_t> MakePackedValidityBuffer(int64_t num_elements) {
+  return std::shared_ptr<uint8_t>(new uint8_t[BitPackedLength(num_elements)]);
```

Comment:
Maybe you want to use the built-in allocation utilities and use `arrow::MemoryPool`? https://github.com/apache/arrow/blob/b496913e5ddb7d8bbef65a59be5a54ac14bf6740/cpp/src/arrow/buffer.h#L402

This has the side benefit of properly capturing allocation errors (which would occur as `std::bad_alloc` here I think)

Reply:
Right, also I'm not sure it's ok to use `shared_ptr` around array allocations.

Reply:
@wesm I changed the code to use `AllocateResizableBuffer` with the default Arrow memory pool since we later use that `uint8_t` array as a back-end for a `MutableBuffer` anyway.

@pitrou that's a good point, yes, I believe that it is undefined behavior to use `std::shared_ptr<T>` as the type for data that has been allocated using `new[]`. This should be fixed in the latest commit by using `AllocateResizableBuffer` instead.